src/browser/new-browser.ts

```diff
@@ -175,6 +204,43 @@ export class NewBrowser extends Browser {
         return assign({}, desiredCapabilities, { [versionKeyName]: this.version });
     }
 
+    protected async _getLocalWebdriverGridUrl(): Promise<string> {
+        if (!this._wdPool) {
+            throw new Error("webdriver pool is not defined");
+        }
+
+        if (this._wdProcess) {
+            return this._wdProcess.gridUrl;
+        }
+
+        this._wdProcess = await this._wdPool.getWebdriver(
+            this._config.desiredCapabilities?.browserName as SupportedBrowser,
+            this._config.desiredCapabilities?.browserVersion as string,
+            { debug: this._config.system.debug },
+        );
+
+        return this._wdProcess.gridUrl;
+    }
+
+    protected async _addExecutablePath(
+        config: BrowserConfig,
+        capabilities: WebdriverIO.Capabilities,
+    ): Promise<WebdriverIO.Capabilities> {
+        const browserNameLowerCase = config.desiredCapabilities?.browserName?.toLowerCase() as string;
+        const executablePath = await installBrowser(
+            this._config.desiredCapabilities?.browserName as SupportedBrowser,
+            this._config.desiredCapabilities?.browserVersion as string,
+        );
```

Comment:
Installs browser (if necessary) and returns its executable path
For safari and edge `null` is returned (we won't set binary path in capabilities, but wdio will be able to find them, if they are installed)